src/rules/switchDefaultRule.ts

```diff
@@ -15,41 +15,216 @@
-        if (
-            node.kind === ts.SyntaxKind.SwitchStatement &&
-            !(node as ts.SwitchStatement).caseBlock.clauses.some(isDefaultClause)
-        ) {
-            ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        if (isSwitchStatement(node) && !node.caseBlock.clauses.some(isDefaultClause)) {
+            if (!ctx.options.allowExhaustive) {
+                ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+            } else if (checker === undefined) {
+                throw Error("Expected `checker` to be defined");
+            } else {
+                const uncovered = uncoveredValues(checker, node);
+
+                if (uncovered === "non-exhaustive") {
+                    ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+                } else if (uncovered.length > 0) {
+                    ctx.addFailureAtNode(node, Rule.failureStringForMissingCases(uncovered));
+                }
+            }
         }
         return ts.forEachChild(node, cb);
     });
 }
+
+// returns a (possibly empty) array of missing values, or the string 'non-exhaustive' if the
+// type of the switched expression can't possibly be matched exhaustively
+function uncoveredValues(
+    checker: ts.TypeChecker,
+    switchStatement: ts.SwitchStatement,
+): "non-exhaustive" | SingletonValue[] {
+    const switchedExprType = checker.getTypeAtLocation(switchStatement.expression);
+
+    const possibleValues: SingletonValue[] = [];
+    for (const type of unionTypeParts(switchedExprType)) {
+        // Note: `boolean` is represented as `true | false` under the hood, so we don't have to
+        // do anything extra to support it
+        const value = singletonValueOfType(type);
+
+        if (value === undefined) {
+            return "non-exhaustive";
+        }
+
+        possibleValues.push(value.value);
+    }
+
+    const caseClauses = switchStatement.caseBlock.clauses as ts.NodeArray<ts.CaseClause>;
+
+    const caseValues: SingletonValue[] = [];
+    for (const clause of caseClauses) {
+        const type = checker.getTypeAtLocation(clause.expression);
+        const value = singletonValueOfType(type);
+
+        if (value === undefined) {
+            return "non-exhaustive";
+        }
+
+        caseValues.push(value.value);
+    }
+
+    // PseudoBigInt is an object, and indexOf will use reference equality. But this should be OK
+    // because AFAICT TypeScript interns all literal types in a Map, and if the type checker returns
+    // the same type twice, it will be the same object (and therefore that object's `value` field
+    // will be the same, too).
+    return possibleValues.filter(value => caseValues.indexOf(value) === -1);
+}
+
+// ts.PseudoBigInt was added in Typescript 3.2, and is a possible `LiteralType` value.
+// Since it is not available in earlier versions, we copy its definition and use it here.
+interface PseudoBigInt {
+    negative: boolean;
+    base10Value: string;
+}
```

Comment:
As far as I can tell, tslint is always compiled using the version of typescript listed under `devDependencies` in package.json. So if we upgrade that dependency to typescript 3.2, I can remove this definition and use `ts.PseudoBigInt` directly.

Reply:
SGTM!